crates/language/src/language_settings.rs

```diff
@@ -35,13 +36,75 @@ pub fn init(cx: &mut AppContext) {
 }
 
 /// Returns the settings for the specified language from the provided file.
-pub fn language_settings<'a>(
+pub fn language_settings(
     language: Option<&Arc<Language>>,
     file: Option<&Arc<dyn File>>,
-    cx: &'a AppContext,
-) -> &'a LanguageSettings {
+    cx: &AppContext,
+) -> LanguageSettings {
     let language_name = language.map(|l| l.name());
-    all_language_settings(file, cx).language(language_name.as_deref())
+    let mut settings = all_language_settings(file, cx)
+        .language(language_name.as_deref())
+        .clone();
+    let path = file
+        .and_then(|f| f.as_local())
+        .map(|f| f.abs_path(cx))
+        .unwrap_or_else(|| std::path::PathBuf::new());
```

Comment:
I do not think it's worth looking for editorconfig in case there's no path — let's extract the whole new code batch into something line `fn editor_config_settings(...) -> Option<LanguageSettingsContent>`.

Note: I'm also somewhat concerned that we do recursive directory to re-read the same file, over and over (instead of caching it) due to `ec4rs::properties_of` API usage.
https://docs.rs/ec4rs/latest/src/ec4rs/file.rs.html#97

I do not insist on this being optimized right now, but feels very good.
It's somewhat complex as we do not have any `self` around though.

Reply:
> Note: I'm also somewhat concerned that we do recursive directory to re-read the same file, over and over (instead of caching it) due to ec4rs::properties_of API usage.

Yep, but not doing that would likely mean rewriting the lib I guess (caching in the `Config::open` method and updating cache if any change is detected). Another issue with this is if a user is abusing the `root` property of editorconfig and having a root in a folder parent of the zed workspace...